### PR TITLE
added merge functionality to javascript client sdk

### PR DESCRIPTION
**Pull request by PJGitLan**

### CODE REVIEW

**Comment by thjaeckle** [comment]
Hello @PJGitLan and thanks for this contribution.
Will take a look once the "Checks" for this PR are all green.

For the "eclipsefdn/eca" check you will have to create an Eclipse account and sign the ECA (Eclipse Contributor Agreement), as also noted in the [CONTRIBUTING.md](https://github.com/eclipse-ditto/ditto-clients/blob/master/CONTRIBUTING.md#legal-requirements) - and sign your commit with the email address registered in your Eclipse account.

File: javascript/lib/api/src/client/request-factory/request-sender.ts [hunk 1]
```diff
@@ -51,6 +51,19 @@ export abstract class RequestSender {
       });
   }
 
+  public fetchMergeRequest<T>(options: ParseRequest<T>): Promise<MergeResponse<T>> {
+    return this.fetchRequest(options).then(response => {
+      if (response.status === 204) {
```
Comment (thjaeckle) [inline javascript/lib/api/src/client/request-factory/request-sender.ts:56]: As far as I remember a PATCH / merge could also create a new feature or property ..
So IMO checking for the status code 201 is missing here.

File: javascript/lib/api/src/client/request-factory/request-sender.ts [hunk 2]
```diff
@@ -51,6 +51,19 @@ export abstract class RequestSender {
...
+        return new MergeResponse(
+          options.parser(response.body),
```
Comment (thjaeckle) [inline javascript/lib/api/src/client/request-factory/request-sender.ts:58]: For status 204 (no content) there is no content in the response, so parsing it is not required, you can pass in `null` (same as for PUT).

File: javascript/lib/api/src/client/request-factory/websocket-request-sender.ts [hunk 3]
```diff
@@ -95,7 +98,8 @@ export class WebSocketRequestSender extends RequestSender {
   public fetchRequest(options: FetchRequest): Promise<GenericResponse> {
     const topic = this.buildTopic(options.id, this.group, 'commands', WebSocketRequestSender.translateVerb(options.verb));
     const path = WebSocketRequestSender.buildPath(options.path);
-    const headers = this.buildHeaders(options.requestOptions, { 'content-type': 'application/json' });
+    const requestOptions = options?.requestOptions;
+    const headers = this.buildHeaders(requestOptions, requestOptions?.getHeaders().has('Content-Type') === true ? undefined : { 'Content-Type': 'application/json' });
```
Comment (thjaeckle) [inline javascript/lib/api/src/client/request-factory/websocket-request-sender.ts:102]: 👍

File: javascript/lib/api/src/model/response.ts [hunk 4]
```diff
@@ -65,6 +65,27 @@ export class PutResponse<T> implements GenericResponse {
...
 }
 
+export class MergeResponse<T> implements GenericResponse {
+  public constructor(
+    private readonly _value: T | null,
+    private readonly _status: number,
+    private readonly _headers: Map<string, string>
+  ) {}
+
+
```
Comment (thjaeckle) [inline javascript/lib/api/src/model/response.ts:75]: `wasCreated` and `wasUpdated` (like for the PutResponse) could also be added.

**Review by thjaeckle — CHANGES_REQUESTED** [review]
Thanks @PJGitLan for the contribution. 

I added some remarks / wishes for changing inline.
Good that you added "some" merge functionality to the JS SDK. 👍 

However, it would be nice if we go into that direction in the JS SDK to support PATCH / merge, that all levels are supported.
Do you think you can also address all other levels (where currently only "PUT" is available in the client)?
From the top of my head:
* merge Thing
* merge Thing attributes
* merge Thing attribute
* merge Features (all)
* merge Feature (single)
* merge Feature desired properties

That way the complete "merge" would be available in the JS client and not only a small selected subpart.

**Comment by PJGitLan** [comment]
> Thanks @PJGitLan for the contribution.
> 
> I added some remarks / wishes for changing inline. Good that you added "some" merge functionality to the JS SDK. +1
> 
> However, it would be nice if we go into that direction in the JS SDK to support PATCH / merge, that all levels are supported. Do you think you can also address all other levels (where currently only "PUT" is available in the client)? From the top of my head:
> 
> * merge Thing
> * merge Thing attributes
> * merge Thing attribute
> * merge Features (all)
> * merge Feature (single)
> * merge Feature desired properties
> 
> That way the complete "merge" would be available in the JS client and not only a small selected subpart.

I added further merge functionality and solved previous remarks

File: javascript/lib/api/src/model/response.ts (hunk 4 above)
Comment (PJGitLan) [reply javascript/lib/api/src/model/response.ts]: Looks like putResponse en mergeResponse are the same so I removed mergeResponse and used putResponse instead.

File: javascript/lib/api/src/client/request-factory/request-sender.ts (hunk 2 above)
Comment (PJGitLan) [reply javascript/lib/api/src/client/request-factory/request-sender.ts]: Went back to using PutResponse

File: javascript/lib/api/src/client/handles/things.interfaces.ts [hunk 5]
```diff
@@ -142,6 +172,17 @@ export interface HttpThingsHandle extends ThingsHandle {
    */
   putPolicyId(thingId: string, policyId: string, options?: MatchOptions): Promise<PutResponse<string>>;
```
Comment (PJGitLan) [inline javascript/lib/api/src/client/handles/things.interfaces.ts:173]: Any reason why policyId and definition are only used for http-requests and not for websocket-request?

File: javascript/lib/api/src/client/request-factory/request-sender.ts (hunk 1 above)
Comment (PJGitLan) [reply javascript/lib/api/src/client/request-factory/request-sender.ts]: this should be resolved by using fetchPutRequest()

File: javascript/lib/api/src/client/constants/http-verb.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
+export enum HttpVerb {
```
Comment (thjaeckle) [inline javascript/lib/api/src/client/constants/http-verb.ts:14]: Nice that you added this enum 👍 
I however would have expected that the existing code is then also adjusted to use the enum. 
That should hopefully, using "search/replace" in the IDE, not be too much work.
Could you please do that?

File: javascript/lib/api/src/client/constants/header.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
+export enum Header {
```
Comment (thjaeckle) [inline javascript/lib/api/src/client/constants/header.ts:14]: Nice that you added this enum 👍
I however would have expected that the existing code is then also adjusted to use the enum.
That should hopefully, using "search/replace" in the IDE, not be too much work.
Could you please do that?

File: javascript/lib/api/src/client/constants/content-type.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
+export enum ContentType {
```
Comment (thjaeckle) [inline javascript/lib/api/src/client/constants/content-type.ts:14]: Nice that you added this enum 👍
I however would have expected that the existing code is then also adjusted to use the enum.
That should hopefully, using "search/replace" in the IDE, not be too much work.
Could you please do that?

**Review by thjaeckle — CHANGES_REQUESTED** [review]
Thanks, that looks very promising 👍 
I added some remarks, mainly about the js-doc, which were copy&pasted without adjustment at some places.

I like the added enums, but we then already should use them everywhere to get the JS client in a better shape slowly :)

File: javascript/lib/api/src/client/handles/things.interfaces.ts [hunk 9]
```diff
@@ -66,6 +66,27 @@ export interface ThingsHandle {
...
+   * @param options - Options to use for the request.
+   * @returns A Promise for the response
+   */
+  patchAttributes(thingId: string, attributes: object, options?: MatchOptions): Promise<PutResponse<object>>;
+
+  /**
+   * merges an Attribute of a Thing.
+   *
+   * @param thingId - The ID of the Thing.
+   * @param attributePath - The new path to the new Attribute.
```
[the remark this answers is not in the document]
Comment (PJGitLan) [reply javascript/lib/api/src/client/handles/things.interfaces.ts]: done

File: javascript/lib/api/src/client/handles/things.interfaces.ts [hunk 10]
```diff
@@ -66,6 +66,27 @@ export interface ThingsHandle {
...
+   * @param attributeValue - The new Attribute.
```
[the remark this answers is not in the document]
Comment (PJGitLan) [reply javascript/lib/api/src/client/handles/things.interfaces.ts]: done

File: javascript/lib/api/src/client/handles/features.interfaces.ts [hunk 11]
```diff
@@ -89,6 +107,16 @@ export interface FeaturesHandle {
    */
   putDefinition(featureId: string, definition: string[], options?: MatchOptions): Promise<PutResponse<string[]>>;
 
+  /**
+   * Merges the definition of the specified Feature.
+   *
+   * @param featureId - The ID of the Feature.
+   * @param definition - The new definition.
```
[the remark this answers is not in the document]
Comment (PJGitLan) [reply javascript/lib/api/src/client/handles/features.interfaces.ts]: done

File: javascript/lib/api/src/client/handles/features.interfaces.ts [hunk 12]
```diff
@@ -110,6 +138,27 @@ export interface FeaturesHandle {
...
+   * @returns A Promise for the response
+   */
+  patchProperties(featureId: string, properties: object, options?: MatchOptions): Promise<PutResponse<Object>>;
+
+  /**
+   * Merge a Property of the specified Feature.
+   *
+   * @param featureId - The ID of the Feature.
+   * @param propertyPath - The path to the Property.
+   * @param property - The Property.
```
[the remark this answers is not in the document]
Comment (PJGitLan) [reply javascript/lib/api/src/client/handles/features.interfaces.ts]: done

File: javascript/lib/api/src/client/handles/things.interfaces.ts [hunk 13]
```diff
@@ -123,6 +144,15 @@ export interface HttpThingsHandle extends ThingsHandle {
    */
   putThing(thing: Thing, options?: MatchOptions): Promise<PutResponse<Thing>>;
 
+  /**
+   * Adds or merges a Thing.
+   *
+   * @param thing - The Thing to create.
```
[the remark this answers is not in the document]
Comment (PJGitLan) [reply javascript/lib/api/src/client/handles/things.interfaces.ts]: done

File: javascript/lib/api/src/client/handles/things.interfaces.ts [hunk 14]
```diff
@@ -142,6 +172,17 @@ export interface HttpThingsHandle extends ThingsHandle {
    */
   putPolicyId(thingId: string, policyId: string, options?: MatchOptions): Promise<PutResponse<string>>;
 
+
+  /**
+   * Adds or updates the policyId of a Thing.
+   *
+   * @param thingId - The ID of the Thing.
+   * @param policyId - The new PolicyId.
```
[the remark this answers is not in the document]
Comment (PJGitLan) [reply javascript/lib/api/src/client/handles/things.interfaces.ts]: done

File: javascript/lib/api/src/client/handles/things.interfaces.ts [hunk 15]
```diff
@@ -66,6 +66,27 @@ export interface ThingsHandle {
    */
   putAttribute(thingId: string, attributePath: string, attributeValue: any, options?: MatchOptions): Promise<PutResponse<any>>;
 
+  /**
+   * Merges the Attributes of a Thing.
+   *
+   * @param thingId - The ID of the Thing.
+   * @param attributes - The new Attributes.
```
[the remark this answers is not in the document]
Comment (PJGitLan) [reply javascript/lib/api/src/client/handles/things.interfaces.ts]: done

File: javascript/lib/api/src/client/handles/features.interfaces.ts [hunk 16]
```diff
@@ -110,6 +138,27 @@ export interface FeaturesHandle {
    */
   putProperty(featureId: string, propertyPath: string, property: any, options?: MatchOptions): Promise<PutResponse<any>>;
 
+  /**
+   * Merges the properties of the specified Feature.
+   *
+   * @param featureId - The ID of the Feature.
+   * @param properties - The properties.
```
[the remark this answers is not in the document]
Comment (PJGitLan) [reply javascript/lib/api/src/client/handles/features.interfaces.ts]: done

File: javascript/lib/api/src/client/handles/things.interfaces.ts [hunk 17]
```diff
@@ -161,6 +202,16 @@ export interface HttpThingsHandle extends ThingsHandle {
    */
   putDefinition(thingId: string, definition: string, options?: MatchOptions): Promise<PutResponse<string>>;
 
+  /**
+   * merges the definition of a Thing.
+   *
+   * @param thingId - The ID of the Thing.
+   * @param definition - The new definition.
```
[the remark this answers is not in the document]
Comment (PJGitLan) [reply javascript/lib/api/src/client/handles/things.interfaces.ts]: done

File: javascript/lib/api/src/client/handles/features.interfaces.ts [hunk 18]
```diff
@@ -70,6 +70,24 @@ export interface FeaturesHandle {
    */
   putFeatures(features: Features, options?: MatchOptions): Promise<PutResponse<Features>>;
 
+  /**
+   * Merges a Feature of this handle's Thing.
+   *
+   * @param feature - The new Feature.
```
[the remark this answers is not in the document]
Comment (PJGitLan) [reply javascript/lib/api/src/client/handles/features.interfaces.ts]: done

File: javascript/lib/api/src/client/handles/features.interfaces.ts [hunk 19]
```diff
@@ -70,6 +70,24 @@ export interface FeaturesHandle {
...
+   * @param options - Options to use for the request.
+   * @returns A Promise for the response
+   */
+  patchFeature(feature: Feature, options?: MatchOptions): Promise<PutResponse<Feature>>;
+
+  /**
+   * Merges all Features of this handle's Thing.
+   *
+   * @param features - The new Features.
```
[the remark this answers is not in the document]
Comment (PJGitLan) [reply javascript/lib/api/src/client/handles/features.interfaces.ts]: done

File: javascript/lib/api/src/client/constants/http-verb.ts (hunk 6 above)
Comment (PJGitLan) [reply javascript/lib/api/src/client/constants/http-verb.ts:14]: No problem, refactored the existing code

File: javascript/lib/api/src/client/constants/header.ts (hunk 7 above)
Comment (PJGitLan) [reply javascript/lib/api/src/client/constants/header.ts:14]: done

File: javascript/lib/api/src/client/constants/content-type.ts (hunk 8 above)
Comment (PJGitLan) [reply javascript/lib/api/src/client/constants/content-type.ts:14]: done

**Review by thjaeckle — APPROVED** [review]
@PJGitLan the changes look good to me, very nice :+1:
Great that you did the extra changes to completely support "merge". Thanks you for the contribution.

**Comment by PJGitLan** [comment]
> @PJGitLan the changes look good to me, very nice +1 Great that you did the extra changes to completely support "merge". Thank you for the contribution.

Hi @thjaeckle, no problem. Any idea when version 3.2.0 will be published?

**Comment by thjaeckle** [comment]
Today 😉

**Pull request merged**